src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeMemory.Unix.cs

```diff
@@ -0,0 +1,126 @@
+        /// </remarks>
+        [CLSCompliant(false)]
+        public static void* AlignedAlloc(nuint alignment, nuint byteCount)
+        {
+            nuint size = (byteCount != 0) ? byteCount : 1;
+            void* result = (void*)(nint)Interop.Sys.MemAlignedAlloc(alignment, size);
+
+            if (result == null)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            return result;
+        }
+
+        /// <summary>Frees an aligned block of memory.</summary>
+        /// <param name="ptr">A pointer to the aligned block of memory that should be freed.</param>
+        /// <remarks>
+        ///    <para>This method does nothing if <paramref name="ptr" /> is <c>null</c>.</para>
+        ///    <para>This method is a thin wrapper over the C <c>free</c> API or a platform dependent aligned free API such as <c>_aligned_free</c> on Win32.</para>
+        /// </remarks>
+        [CLSCompliant(false)]
+        public static void AlignedFree(void* ptr)
+        {
+            Interop.Sys.MemAlignedFree((nint)ptr);
+        }
+
+        /// <summary>Allocates a block of memory of the specified size, in bytes.</summary>
+        /// <param name="byteCount">The size, in bytes, of the block to allocate.</param>
+        /// <returns>A pointer to the allocated block of memory.</returns>
+        /// <exception cref="OutOfMemoryException">Allocating <paramref name="byteCount" /> of memory failed.</exception>
+        /// <remarks>This method is a thin wrapper over the C <c>malloc</c> API.</remarks>
+        [CLSCompliant(false)]
+        public static void* Alloc(nuint byteCount)
+        {
+            nuint size = (byteCount != 0) ? byteCount : 1;
+            void* result = (void*)(nint)Interop.Sys.MemAlloc(size);
+
+            if (result == null)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            return result;
+        }
+
+        /// <summary>Allocates and zeroes a block of memory of the specified size, in elements.</summary>
+        /// <param name="elementCount">The count, in elements, of the block to allocate.</param>
+        /// <param name="elementSize">The size, in bytes, of each element in the allocation.</param>
+        /// <returns>A pointer to the allocated and zeroed block of memory.</returns>
+        /// <exception cref="OutOfMemoryException">Allocating <paramref name="elementCount" /> * <paramref name="elementSize" /> bytes of memory failed.</exception>
+        /// <remarks>
+        ///     <para>This method converts a <paramref name="elementCount" /> or <paramref name="elementSize" /> of <c>0</c> to <c>1</c>.</para>
+        ///     <para>This method is a thin wrapper over the C <c>calloc</c> API.</para>
+        /// </remarks>
+        [CLSCompliant(false)]
+        public static void* AllocZeroed(nuint elementCount, nuint elementSize)
+        {
+            nuint num = (elementCount != 0) ? elementCount : 1;
```

Comment:
This can allocate more than necessary for the 0 case.

Would it be better to do `((elementSize == 0) || (elementCount ==0)) ? Interop.Sys.MemAlloc(1) : Interop.Sys.MemAllocZeroed(...)`?

Reply:
I think this falls under what I commented here: https://github.com/dotnet/runtime/pull/54006#discussion_r649406314

Bumping this to `1` doesn't seem like the "right" behavior to me. I think instead it would make sense to handle `0` as if the underlying API returned `nullptr` (which is what is typically done for "invalid" parameters) and to just throw `OOM`.

I think this is the "least surprising" and helps encourage users to pass in valid inputs, rather than silently do 1-byte allocations (which you almost certainly don't want, since `stackalloc` is better in that case)

Reply:
Will default to `size == 1` as per https://github.com/dotnet/runtime/pull/54006#discussion_r649439220

> This can allocate more than necessary for the 0 case.

Shouldn't this be `1*1 == 1` and so it would be the "same" as `MemAlloc(1)`, but with the memory being zeroed?

Reply:
`AllocZeroed(0, sizeof(BigStructure))`

Reply:
Ah I see now. I don't have a preference so I'll update to do `MemAlloc(1)` as you suggested here